crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -2517,6 +2523,118 @@ pub(crate) mod tests {
+            {"delivery":{"log":[{"time":"2025-03-05T08:49:24.986Z","status":"PENDING"},{"time":"2025-03-05T08:49:25.000Z","status":"SEND"},{"time":"2025-03-05T08:49:25.162Z","status":"DELIVERED"}],"time":"2025-03-05T08:49:25.707Z","status":"PENDING"},"agentId":"1916574062","creationTime":"2025-03-05T08:49:24.967Z","deviceId":"1916574062","id":"16574089","status":"PENDING","c8y_Restart":{},"description":"do something","externalSource":{"externalId":"test-device","type":"c8y_Serial"}}
+        ).to_string());
+        assert_eq!(
+            converter
+                .parse_c8y_devicecontrol_topic(&original)
+                .await
+                .unwrap(),
+            vec![MqttMessage {
+                topic: Topic {
+                    name: "te/device/main///cmd/restart/c8y-mapper-16574089".into(),
+                },
+                payload: json!({"status":"init"}).to_string().into(),
+                qos: QoS::AtLeastOnce,
+                retain: true,
+            },]
+        );
+        assert_eq!(
+            converter
+                .parse_c8y_devicecontrol_topic(&redelivery)
+                .await
+                .unwrap(),
+            vec![]
+        );
+    }
+
+    #[tokio::test]
+    async fn custom_operations_are_deduplicated() {
+        let tmp_dir = TempTedgeDir::new();
+        let custom_op = r#"exec.topic = "my/custom/topic"
+        exec.on_fragment = "my_op"
+        exec.command = "/etc/tedge/operations/command ${.payload.my_op.text}""#;
+        let f = tmp_dir.dir("operations").dir("c8y").file("my_op");
+        f.with_raw_content(custom_op);
+        let (mut converter, _http_proxy) = create_c8y_converter(&tmp_dir).await;
+
+        let original = MqttMessage::new(&Topic::new_unchecked("my/custom/topic"), json!(
+            {"delivery":{"log":[],"time":"2025-03-05T08:49:24.986Z","status":"PENDING"},"agentId":"1916574062","creationTime":"2025-03-05T08:49:24.967Z","deviceId":"1916574062","id":"16574089","status":"PENDING","my_op":{},"description":"do something","externalSource":{"externalId":"test-device","type":"c8y_Serial"}}
+        ).to_string());
+        let redelivery = MqttMessage::new(&Topic::new_unchecked("my/custom/topic"), json!(
+            {"delivery":{"log":[{"time":"2025-03-05T08:49:24.986Z","status":"PENDING"},{"time":"2025-03-05T08:49:25.000Z","status":"SEND"},{"time":"2025-03-05T08:49:25.162Z","status":"DELIVERED"}],"time":"2025-03-05T08:49:25.707Z","status":"PENDING"},"agentId":"1916574062","creationTime":"2025-03-05T08:49:24.967Z","deviceId":"1916574062","id":"16574089","status":"PENDING","my_op":{},"description":"do something","externalSource":{"externalId":"test-device","type":"c8y_Serial"}}
+        ).to_string());
+        assert_ne!(
+            converter
+                .parse_json_custom_operation_topic(&original)
+                .await
+                .unwrap(),
+            vec![],
+            "Initial operation delivery produces outgoing message"
+        );
```

Comment:
A slightly stricter check that validates at least the `cmd` topic would have been better than this non-empty check. To avoid false positives like a converted error message sent to `te/errors` instead of the expected `cmd` to beat this assertion.